imap_processing/hi/l1a/science_direct_event.py

```diff
@@ -0,0 +1,288 @@
+        # respectively. This is used to identify which detector
+        # was hit first for this current direct event.
+        trigger_id.append(event["start_bitmask_data"])
+        tof_1.append(event["tof_1"])
+        tof_2.append(event["tof_2"])
+        tof_3.append(event["tof_3"])
+        de_tag.append(event["de_tag"])
+
+    # calculate direct event time using time information from metaevent
+    # and de_tag. epoch in this dataset is the time of the event
+    epoch_datetime = [
+        get_direct_event_time(met_seconds[i], met_subseconds[i], de_tag[i])
+        for i in range(len(met_seconds))
+    ]
+    epoch_time = xr.DataArray(
+        epoch_datetime,
+        name="epoch",
+        dims=["epoch"],
+        attrs=ConstantCoordinates.EPOCH,
+    )
+
+    dataset = xr.Dataset(
+        coords={"epoch": epoch_time},
+        attrs=hi_cdf_attrs.hi_de_l1a_attrs.output(),
+    )
+
+    dataset["met_subseconds"] = xr.DataArray(
+        met_subseconds, dims="epoch", attrs=hi_cdf_attrs.met_subseconds_attrs.output()
+    )
+    dataset["met_seconds"] = xr.DataArray(
+        met_seconds, dims="epoch", attrs=hi_cdf_attrs.met_seconds_attrs.output()
+    )
+    dataset["esa_step"] = xr.DataArray(
+        esa_step, dims="epoch", attrs=hi_cdf_attrs.esa_step_attrs.output()
+    )
+    dataset["trigger_id"] = xr.DataArray(
+        trigger_id, dims="epoch", attrs=hi_cdf_attrs.trigger_id_attrs.output()
+    )
+    dataset["tof_1"] = xr.DataArray(
+        tof_1, dims="epoch", attrs=hi_cdf_attrs.tof_attrs.output()
+    )
+    dataset["tof_2"] = xr.DataArray(
+        tof_2, dims="epoch", attrs=hi_cdf_attrs.tof_attrs.output()
+    )
+    dataset["tof_3"] = xr.DataArray(
+        tof_3, dims="epoch", attrs=hi_cdf_attrs.tof_attrs.output()
+    )
+    dataset["de_tag"] = xr.DataArray(
+        de_tag, dims="epoch", attrs=hi_cdf_attrs.de_tag_attrs.output()
+    )
+    # TODO: figure out how to store information about
+    # input data(one or more) it used to produce this dataset
+    return dataset
+
+
+def science_direct_event(packets_data: list):
+    """Unpack IMAP-Hi direct event data.
+
+    Processing step:
+        1. Break binary stream data into unit of 48-bits
+        2. Parse direct event data
+        5. Save the data into xarray dataset.
+    """
+    event_48bits_list = []
+    # TODO: ask Paul if he wants MET time in the dataset
+    # If so, get MET time of every packets
+
+    # Because DE_TOF is a variable length data,
+    # I am using extend to add another list to the
+    # end of the list. This way, I don't need to flatten
+    # the list later.
+    for data in packets_data:
+        # break binary stream data into unit of 48-bits
+        event_48bits_list.extend(break_into_bits_size(data.data["DE_TOF"].raw_value))
+
+    # Parse 48-bits into meaningful data such as metaevent or direct event
+    de_data_list = [parse_direct_event(event) for event in event_48bits_list]
```

Comment:
Could you move this up into the for-loop to avoid looping through twice? (`de_data_list.append(parse_direct_event(event_48bits_list.extend(break_into_bits_size(data.data["DE_TOF"].raw_value))))`)

Reply:
yes. for sure! I totally missed that.